### DISCUSSION TITLE
Optimized JSON Deserialization

### DISCUSSION THREAD

**Discussion by tkauf15k** [sqlalchemy/sqlalchemy#9040] [Usage Questions]
Hi, 

i'm using sqlalchemy with postgres and for a couple of use-cases JSON columns as well. In addition, i make use of pydantic for specifying DTOs for the objects (that i then store in the json columns). At the moment i basically convert my objects to/from dicts at the application layer, but this leads to quite a lot of boilerplate code so i thought about some alternative and most importantly also more efficient solutions.

In a first step, I tried to override the serializer/deserializer, but since it basically follows the interface of json.dumps/loads, i do not really know to what type to deserialize my json to. I played a little bit around with some pydantic features, e.g. wrapping all DTO types inside a union in a container and let pydantic find the correct one, but as the number of dto-types and jsons in the result set increases this is apparently highly inefficient. 

In a next step, i took a look at the JSON type itself and at first glance it seemed promising to simply override the `bind_processor `method in custom subclass; however i found that deserialization is not conducted in the respective `result_processor` inside the type but rather already in psycopg2 and even more i the json-type itself does not have any means to access my dto-type information (e.g. passed by kwargs etc.) 

From a user perspective i'd envision something like specifying a type inside the column (*or* maybe even just a type annotation, although that's probaly tricky and too implicit) 

`settings: MySettings = Column(JSON, type=MySettings)`

that is then passed as argument to the custom serializer/deserialzer. 

Btw, i also considered dynamically adding type information to my jsons (like putting them inside a container like {"$type":"Settings", "data":...}, but somewhat i have the feeling that it polutes my db design as i would pass information through the DB that is already available in my schema. even more it would require me again to perform an additional step to obtain the type before deserializing into my pydantic object...

Anyone here that has an idea how to alternatively approach this?

kind regards!

**Comment by CaselIT** [comment]
Hi,

I think you can create a type decorator for this.
Take a look at this reply, where one is used for path like object, but it should adapt very easily to your use use case:
https://github.com/sqlalchemy/sqlalchemy/discussions/9027#discussioncomment-4510017
